server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java

```diff
@@ -0,0 +1,130 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+import org.apache.james.mailrepository.api.MailRepositoryStore;
+import org.apache.james.metrics.logger.DefaultMetricFactory;
+import org.apache.james.webadmin.WebAdminServer;
+import org.apache.james.webadmin.WebAdminUtils;
+import org.apache.james.webadmin.service.MailRepositoryStoreService;
+import org.apache.james.webadmin.utils.JsonTransformer;
+import org.eclipse.jetty.http.HttpStatus;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableList;
+import com.jayway.restassured.RestAssured;
+import com.jayway.restassured.builder.RequestSpecBuilder;
+import com.jayway.restassured.http.ContentType;
+
+public class MailRepositoriesRoutesTest {
+
+    private WebAdminServer webAdminServer;
+    private MailRepositoryStore mailRepositoryStore;
+
+    @Before
+    public void setUp() throws Exception {
+        mailRepositoryStore = mock(MailRepositoryStore.class);
+
+        webAdminServer = WebAdminUtils.createWebAdminServer(
+                new DefaultMetricFactory(),
+                new MailRepositoriesRoutes(new MailRepositoryStoreService(mailRepositoryStore), new JsonTransformer()));
+        webAdminServer.configure(NO_CONFIGURATION);
+        webAdminServer.await();
+
+        RestAssured.requestSpecification = new RequestSpecBuilder()
+            .setContentType(ContentType.JSON)
+            .setAccept(ContentType.JSON)
+            .setBasePath(MailRepositoriesRoutes.MAIL_REPOSITORIES)
+            .setPort(webAdminServer.getPort().get().getValue())
+            .setConfig(newConfig().encoderConfig(encoderConfig().defaultContentCharset(StandardCharsets.UTF_8)))
+            .build();
+    }
+
+    @After
+    public void tearDown() {
+        webAdminServer.destroy();
+    }
+
+    @Test
+    public void getMailRepositoriesShouldReturnEmptyWhenEmpty() {
+        List<Object> mailRepositories =
+            when()
+                .get()
+            .then()
+                .statusCode(HttpStatus.OK_200)
+                .contentType(ContentType.JSON)
+                .extract()
+                .body()
+                .jsonPath()
+                .getList(".");
+
+        assertThat(mailRepositories).isEmpty();
+    }
+
+    @Test
+    public void getMailRepositoriesShouldReturnRepositoryWhenOne() {
+        when(mailRepositoryStore.getUrls())
+            .thenReturn(ImmutableList.of("url://myRepo"));
+
+        when()
+            .get()
+        .then()
+            .statusCode(HttpStatus.OK_200)
+            .body("", hasSize(1))
+            .body("[0].repository", is("url://myRepo"))
+            .body("[0].encodedUrl", is("url%3A%2F%2FmyRepo"));
```

Comment:
I don't understand why we need to return that encoded version, I mean, the consumer is probably able to encode an URL ...

Reply:
ideally we should return a link, but I did not find a good way for that. I find it convenient anyway, just get this value and use it in after

Reply:
there's a problem in the API. Either it's an opaque value and we call it "id" or it's an URL and you should not encode it + you should be able to access it.
IMO, it's an ID.

Reply:
done, but that's the type of comments I would like to have earlier